src/sentry/models/outbox.py

```diff
@@ -593,9 +593,18 @@ def process_coalesced(
         # If the context block didn't raise we mark messages as completed by deleting them.
         if coalesced is not None:
             assert first_coalesced, "first_coalesced incorrectly set for non-empty coalesce group"
-            deleted_count, _ = (
-                self.select_coalesced_messages().filter(id__lte=coalesced.id).delete()
-            )
+            deleted_count = 0
+
+            # Use a fetch and delete loop as doing cleanup in a single query
+            # causes timeouts with large datasets. Apply the ID condition in
+            # python as filtering rows in postgres leads to timeouts
+            delete_ids = []
+            for item in self.select_coalesced_messages()[:100]:
```

Comment:
I think this is different than @corps' solution. Instead of limiting to 100 messages in the coalesced group, his proposed code iterates through all coalesced messages in batches of 100, then iterates through each batch, filters all of the relevant IDs and then constructs a filter query for each batch.

While this will help delete performance, won't this also potentially introduce other issues for coalesced groups > 100 messages?

Reply:
Correct.  The idea is to process in discrete batches of 100, but to still attempt to process everything that can be.  The performance issue is the fact that all concurrent queries, in attempting to sort by id (non indexed) are simulat inflicting that performance on the db.  but in the case of concurrent processes working discrete batches and *not* forcing the search, there is cooperation that results in how each part is reducing the space of work to be done on the db side, too.

Reply:
Ok, I missed that intent, as the suggested code doesn't fetch in batches and I assumed the pseudocode was incomplete. I can add batched iteration.